Mathlib/Combinatorics/SimpleGraph/Tutte.lean

```diff
@@ -5,31 +5,70 @@ Authors: Pim Otte
-  perfect matching.
--/
+* `SimpleGraph.TutteViolator G u` is a set of vertices `u` such that the amount of
+  odd components left after deleting `u` from `G` is larger than the number of vertices in `u`.
+  This certificies non-existence of a perfect matching.
 
+## Main results
+* `SimpleGraph.tutte` states Tutte's theorem: A graph has a perfect matching, if and
+  only if no Tutte violators exist.
+-/
 namespace SimpleGraph
 
 universe u
-variable {V : Type u} {G G' : SimpleGraph V} {u x v' w : V} [Fintype V]
+variable {V : Type u} {G G' : SimpleGraph V} {u x v' w : V}
 
-/-- A set certifying non-existence of a perfect matching. -/
+/-- A set certifying non-existence of a perfect matching -/
 def IsTutteViolator (G : SimpleGraph V) (u : Set V) : Prop :=
   u.ncard < ((⊤ : G.Subgraph).deleteVerts u).coe.oddComponents.ncard
 
+
+/-- This lemma shows an alternating cycle exists in a specific subcase of the proof
+of Tutte's theorem -/
+private lemma tutte_exists_IsAlternating_IsCycles {x b a c : V} {M : Subgraph (G ⊔ edge a c)}
+    (hM2 : M.IsPerfectMatching) (p : G'.Walk a x) (hp : p.IsPath)
+    (hcalt : G'.IsAlternating M.spanningCoe) (hM2nadj : ¬M.Adj x a)
+    (hpac : p.toSubgraph.Adj a c) (hnpxb : ¬p.toSubgraph.Adj x b) (hM2ac : M.Adj a c)
+    (hgadj : G.Adj x a) (hnxc : x ≠ c) (hnab : a ≠ b)
+    (hle : p.toSubgraph.spanningCoe ≤ G ⊔ edge a c)
+    (aux : (c' : V) → c' ≠ a → p.toSubgraph.Adj c' x → M.Adj c' x) :
+    ∃ G', G'.IsAlternating M.spanningCoe ∧ G'.IsCycles ∧
+    ¬G'.Adj x b ∧ G'.Adj a c ∧ G' ≤ G ⊔ edge a c := by
+  use p.toSubgraph.spanningCoe ⊔ edge x a
+  refine ⟨IsAlternating.sup_edge (hcalt.spanningCoe p.toSubgraph) (by simp_all)
+    (fun u' hu'x hadj ↦ by simpa [← hp.snd_of_toSubgraph_adj hadj, hp.snd_of_toSubgraph_adj hpac])
+      (fun c' hc'a hadj ↦ aux _ hc'a hadj), ?_⟩
+  refine ⟨by
+    refine fun v hv ↦ hp.isCycles_spanningCoe_toSubgraph_sup_edge hgadj.ne.symm (fun hadj ↦ ?_) hv
+    rw [← Walk.mem_edges_toSubgraph, Subgraph.mem_edgeSet] at hadj
+    simp [← hp.snd_of_toSubgraph_adj hadj.symm, hp.snd_of_toSubgraph_adj hpac] at hnxc, ?_⟩
+  exact ⟨by
+    simp only [sup_adj, Subgraph.spanningCoe_adj, hnpxb, edge_adj]; aesop, by aesop,
+    sup_le_iff.mpr ⟨hle, fun v w hvw ↦ by
+      simpa [sup_adj, edge_adj,
+        adj_congr_of_sym2 _ ((adj_edge _ _).mp hvw).1.symm] using .inl hgadj⟩⟩
+
+variable [Fintype V]
+
+lemma IsTutteViolator.mono {u : Set V} (h : G ≤ G') (ht : G'.IsTutteViolator u) :
+    G.IsTutteViolator u := by
+  simp only [IsTutteViolator, Subgraph.induce_verts, Subgraph.verts_top] at *
+  have := ncard_oddComponents_mono _ (Subgraph.deleteVerts_mono' (G := G) (G' := G') u h)
+  simp only [oddComponents] at *
+  omega
+
 /-- Given a graph in which the universal vertices do not violate Tutte's condition,
 if the graph decomposes into cliques, there exists a matching that covers
 everything except some universal vertices.
-
```

Comment:
I would think it is better with the line break

Reply:
Re-added